### DISCUSSION TITLE
Unable to read DHT22 sensor

### DISCUSSION THREAD

**Discussion by aburnett397** [Pi4J/pi4j#191] [Q&A] [answered]
I'm trying to port old pi4j-v1 code to the new API. Reading a DHT22 involves using a pin alternatively as an input and an output. When reading input, the duration between the edges encodes the desired information with the need to differentiate between 70us and 27us pulses.

I am loosely following this [pigpio python implementation](https://github.com/joan2937/pigpio/blob/master/EXAMPLES/Python/DHT22_AM2302_SENSOR/DHT22.py)

I create the input and output for the pin like so:

```
cfgOut = DigitalOutput.newConfigBuilder(pi4j)
  .address(pin)
  .initial(DigitalState.HIGH)
  .shutdown(DigitalState.HIGH)
  .provider("pigpio-digital-output")
  .build();
cfgIn = DigitalInput.newConfigBuilder(pi4j)
  .address(pin)
  .pull(PullResistance.OFF)
  .provider("pigpio-digital-input")
  .build();
```

A reading is then triggered with the pin as an output before switching to an input and registering a listener (this is run every 30 seconds so the created input and output are recycled with `initialize()`):

```
//switch to output
if (out == null)
  out = pi4j.create(cfgOut);
else out.initialize(pi4j);

//set pin LOW for 10ms, then HIGH to trigger response
out.state(DigitalState.LOW);
long now = System.nanoTime();
while (System.nanoTime()-now < 10000000);
out.state(DigitalState.HIGH);
out.shutdown(pi4j);

//switch to input and register listener
if (in == null)
  in = pi4j.create(cfgIn);
else in.initialize(pi4j);
in.addListener(listener);
```

The listener simply tallies up the edges and logs the `System.nanoTime()`.
After a few seconds, well after the DHT22 response time, reading is stopped:

```
in.removeListener(listener);
in.shutdown(pi4j);
```

When set as an input, I am only able to detect from 0 to 2 edges, a far cry from the 40 that the chip emits. And the spacing between those edges is measured as ~3ms up to ~15ms, an order of magnitude longer than what is expected.
Unless my approach is fundamentally wrong, i get the feeling that I am nowhere near the proper timing requirements. For starters, the characteristics of the chip mean that a pin must reconfigure from output to input within 20us.
Does pi4j expose the pigpio bindings in any way? It would be interesting to see the behavior of a1:1 port of the python code.

**Comment by FDelporte** [comment]
In the CrowPi project, students also tried many different approaches to read correct values from the DHT22. But it seems this type of component is not well-suited to be used in combination with an operating system, compared to a realtime system or microcontroller.

They found a solution by using a Linux system driver as you can read here: https://github.com/Pi4J/pi4j-example-crowpi/blob/main/src/main/java/com/pi4j/crowpi/components/HumiTempComponent.java

**Comment by FDelporte** [comment]
There is also an open issue regarding multipurpose GPIO: https://github.com/Pi4J/pi4j-v2/issues/26

**Comment by laemm-line** [comment]
Possible workaround (for timing issues): If not occupied otherwise, maybe the SPI pins can be misappropriated for such purposes.
By using the SPI clock speed, the timing is not messed up by Java's GC pauses or other delays.

I'm doing something similar to control WS2812B LEDs which require timings in the range of a few hundred nano seconds.
Only I'm just using the MOSI pin for data output, but I suppose you could also use the MISO pin for high-frequency input reading.

**Comment by aburnett397** [comment, marked as answer]
Turns out it works well enough by ditching the edge listeners for active polling. You only need to do this for a few ms on the DHT22 so GC pauses are no big issue. The hardest part remains the delay to reconfigure the pin which often exceeds the 40-80us window meaning some bits can be lost. But you can always repeat the operation on a loop until you get a clean read. 
For anyone interested I used this instead of listeners and got a similar result to what I had with wiringPi:

```
		now = System.nanoTime();
		DigitalState state = in.state();
		long val = 0, lastHi = now;
		int read = 0;
		//active polling for 10ms (5.5ms is enough according to datasheet)
		while (System.nanoTime()-now < 10000000)
		{
			DigitalState next = in.state();
			//edge detection
			if (state != next)
			{
				//if this is the beginning of a high interval
				if (next == DigitalState.HIGH)
					lastHi = System.nanoTime();
				//otherwise end of interval so measure length and store bit
				else
				{
					val = (val << 1);
					read++;
					//if bit is 1
					if ((System.nanoTime()-lastHi)/1000 > 48)
						val++;
				}
				state = next;
			}
		}
		in.shutdown(pi4j);
		
		//should be 40 but the first few bits are often missed and often equal 0
		if (read >= 38)
		{
			int hi = (int)((val & 0xff00000000L) >> 32), hd = (int)((val & 0xff000000L) >> 24),
				ti = (int)((val & 0xff0000) >> 16), td = (int)((val & 0xff00) >> 8), 
				cs = (int)(val & 0xff);
			//checksum
			if (cs == ((hi+hd+ti+td) & 0xff))
			{
				double temperature = ((((ti & 0x7f) << 8)+td)/10.)*((ti & 0x80) != 0 ? -1 : 1);
				double humidity = ((hi << 8)+hd)/10.;
				...
			}
		}
```

**Comment by ParanoidUser** [comment]
@aburnett397, do you mind sharing the complete example of it?

I'm trying to follow your steps here by reading data from AM2302 (same as DHT22). And this trivial "hello PI world" task turned into an endless number of trials and errors that end up constantly receiving zero values. 

I've also tried the [HumiTempComponent.java](https://github.com/Pi4J/pi4j-example-crowpi/blob/main/src/main/java/com/pi4j/crowpi/components/HumiTempComponent.java) suggested by @FDelporte, but it also giving me zeros back. At the same time, these measurements work pretty well through Python's [Adafruit_DHT](https://pypi.org/project/Adafruit-DHT) project. 

Here is what I've got: Debian GNU/Linux 11 (bullseye) / openjdk 11.0.15+10 / Pi4J 2.1.1 / AM2302 with data pin connected to PI's Physical Pin 7 (GPIO4).

**Reply by aburnett397** [reply]
@ParanoidUser Here is the code I ended up using. It is pretty much the same as before with some extra config and, crucially, the loop that tries to get a good read from the sensor 40 times before giving up. Even with that many iterations, it will still fail to read properly once in every ~20 tries.

```
import com.pi4j.context.Context;
import com.pi4j.io.gpio.digital.DigitalInput;
import com.pi4j.io.gpio.digital.DigitalInputConfig;
import com.pi4j.io.gpio.digital.DigitalOutput;
import com.pi4j.io.gpio.digital.DigitalOutputConfig;
import com.pi4j.io.gpio.digital.DigitalState;
import com.pi4j.io.gpio.digital.PullResistance;

public class DHT22
{
	public double temperature, humidity;
	
	public DHT22()
	{
	}
	
	public void update()
	{
		double [] res;
		for (int i=0;i<40;i++)
			if ((res = read()) != null)
			{
				temperature = res[0];
				humidity = res[1];
				return;
			}
			else try {Thread.sleep(10);} catch (Exception e) {}
	}
	
	private static Context pi4j = null;
	private void init()
	{
		if (pi4j == null)
			pi4j = Pi4J.newAutoContext();
		
		cfgOut = DigitalOutput.newConfigBuilder(pi4j)
			.address(dht22.pin)
			.initial(DigitalState.HIGH)
			.shutdown(DigitalState.HIGH)
			.provider("pigpio-digital-output")
			.build();
		cfgIn = DigitalInput.newConfigBuilder(pi4j)
			.address(dht22.pin)
			.pull(PullResistance.OFF)
			.provider("pigpio-digital-input")
			.build();
	}
	private DigitalInputConfig cfgIn = null;
	private DigitalOutputConfig cfgOut = null;
	private DigitalOutput out = null;
	private DigitalInput in = null;
	private double [] read()
	{
		if (cfgIn == null)
			init();

		if (out == null)
			out = pi4j.create(cfgOut);
		else out.initialize(pi4j);
		out.state(DigitalState.LOW);
		long now = System.nanoTime();
		while (System.nanoTime()-now < 2000000);
		out.state(DigitalState.HIGH);
		out.shutdown(pi4j);
		
		if (in == null)
			in = pi4j.create(cfgIn);
		else in.initialize(pi4j);
		
		now = System.nanoTime();
		DigitalState state = in.state();
		long val = 0, lastHi = now;
		int read = 0;
		while (System.nanoTime()-now < 10000000)
		{
			DigitalState next = in.state();
			if (state != next)
			{
				if (next == DigitalState.HIGH)
					lastHi = System.nanoTime();
				else
				{
					val = (val << 1);
					read++;
					if ((System.nanoTime()-lastHi)/1000 > 48)
						val++;
				}
				state = next;
			}
		}
		in.shutdown(pi4j);
		
		//should be 40 but the first few bits are often missed and often equal 0
		if (read >= 38)
		{
			int hi = (int)((val & 0xff00000000L) >> 32), hd = (int)((val & 0xff000000L) >> 24),
				ti = (int)((val & 0xff0000) >> 16), td = (int)((val & 0xff00) >> 8), 
				cs = (int)(val & 0xff);
			//checksum
			if (cs == ((hi+hd+ti+td) & 0xff))
			{
				double temperature = ((((ti & 0x7f) << 8)+td)/10.)*((ti & 0x80) != 0 ? -1 : 1);
				double humidity = ((hi << 8)+hd)/10.;
				return new double [] {temperature, humidity};
			}
		}
		return null;
	}
}
```

**Comment by ParanoidUser** [comment]
Works like magic! Thanks a lot, @aburnett397 for sharing that! 🙏
